### ISSUE TITLE
[numpy compat] Tensor incompatible with numpy.random.shuffle

### ISSUE THREAD

**Issue by cdluminate** [pytorch/pytorch#50880]
## 🐛 Bug

Numpy wont correctly shuffle a torch tensor.

## To Reproduce

```python
In [23]: import torch as th                                                     

In [24]: import numpy as np                                                     

In [25]: x = th.arange(10)                                                      

In [26]: x                                                                      
Out[26]: tensor([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [27]: np.random.shuffle(x)                                                   

In [28]: x                                                                      
Out[28]: tensor([0, 1, 2, 3, 4, 5, 0, 4, 8, 2])
```

There are duplicated items in the shuffled result, which is clearly incorrect.

## Environment

```
In [29]: np.__version__                                                         
Out[29]: '1.18.5'

In [30]: th.__version__                                                         
Out[30]: '1.7.0'
```

 - PyTorch Version (e.g., 1.0): 1.7.0
 - OS (e.g., Linux): Linux (debian unstable)
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Build command you used (if compiling from source): N/A
 - Python version: 3.8.3
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

## Additional context

cc @mruberry @rgommers @heitorschueroff

**Comment by rgommers** [comment]
That's a very weird bug. Pretty sure it's a numpy bug though; the issue should be in this odd bit of code:
```
        else:
            # Untyped path.
            with self.lock:
                for i in reversed(range(1, n)):
                    j = random_interval(&self._bitgen, i)
                    x[i], x[j] = x[j], x[i]
```
in https://github.com/numpy/numpy/blob/master/numpy/random/mtrand.pyx#L4389. The non-legacy shuffle API has the same bug:
```
In [3]: rng = np.random.default_rng()                                       

In [4]: import torch                                                        

In [5]: x = torch.arange(10)                                                

In [6]: rng.shuffle(x)                                                      

In [7]: x                                                                   
Out[7]: tensor([0, 0, 2, 3, 3, 1, 0, 1, 2, 1])
```
Can you please open a NumPy issue instead @cdluminate?

**Comment by cdluminate** [comment]
@rgommers also filed a bug against numpy, see https://github.com/numpy/numpy/issues/18206

**Comment by rkern** [comment]
Yeah, it's expecting that objects that fall down to that branch _don't_ have view-based semantics for those indexing operations. There used to be fewer objects with view-based semantics. We take care of the known view-based-semantics for the common use case of multidimensional `ndarray`s in the previous branch. But to do so, we need to rely on `ndarray`-specific operations. I fear that we would need to rely on `tensor`-specific operations to (a) detect that something different should be done and (b) actually perform the shuffle correctly.

**Comment by rkern** [comment]
I'll note that the stdlib `random.shuffle()` has the same issue (it does the same kind of straightforward tuple-swap that we do):

```
[~]
|13> x = th.arange(10)

[~]
|14> import random

[~]
|15> r = random.Random(1234567890)

[~]
|16> r.shuffle(x)

[~]
|17> x
tensor([0, 1, 0, 2, 2, 4, 1, 7, 2, 9])
```

**Comment by rgommers** [comment]
Here's a discussion on a `torch`-based solution: https://discuss.pytorch.org/t/shuffling-a-tensor/25422.

> I fear that we would need to rely on `tensor`-specific operations to (a) detect that something different should be done and (b) actually perform the shuffle correctly.

Which I think means we don't want to do this in numpy, and the code currently makes incorrect assumptions - it should just raise instead? Changing accepted input `array_like` to `ndarray` seems like the way to go, because there's no way to write such code that's valid for an arbitrary object numpy doesn't know anything about.

**Comment by rkern** [comment]
It needs to work with at least `list` as well, which is the main reason that final branch exists. And I'm sure that we're going to get complaints from people who have been happily passing other kinds of mutable sequences (that don't have the unusual view semantics). I would rather document the limitations (shared by `random.shuffle()`) than use overly-restrictive type checks.

**Comment by rgommers** [comment]
Hmm, I'm not sure I agree. For one because these are not the expected semantics for `array_like` (those would normally go through `asarray`), for another because silently incorrect behaviour is way worse than a clear error.





> It needs to work with at least `list` as well, which is the main reason that final branch exists. And I'm sure that we're going to get complaints from people who have been happily passing other kinds of mutable sequences

Why not just add an `if not isinstance(x, collections.Sequence)` check? That's what the docs state, and is something well-defined. Most other non-numpy array and tensor objects aren't going to work either, and users are much more likely to do that than some non-array non-sequence custom object.

**Comment by rkern** [comment]
I wouldn't read too much into the use of "sequence" in the docstring. That dates to long before the `collections.Sequence` ABC was formalized. I'd probably issue a warning if the value isn't a `collections.Sequence`, now that we are seeing a profusion of new array types that may or may not have view semantics. I think even if we eventually want to just `raise`, we'd want a period of warnings first. A warning would help us see who's going to be annoyed by this. I'm sure someone out there has code that (correctly) shuffles an `index`-less `pandas.Series`.

I also wouldn't read too much into the specification of `array_like` in the `Parameters` section as implying that it ought to be converted to an `ndarray`. Someone decided to document this function using `array_like` because they were using `array_like` simply to describe the set of acceptable types (to include at least `ndarray`s and `list`s), not because they meant that extra implication. This dated from the days of the doc-wiki, so I can't trace the contribution any farther. This method is definitely meant to shuffle `list` objects in-place; that was an explicit feature request and explicitly mentioned in the parameter description. If you think that the use of `array_like` necessarily implies conversion, then that's a documentation bug.

**Comment by rgommers** [comment]
Adding a warning for unknown types now sounds good.

> If you think that the use of `array_like` necessarily implies conversion, then that's a documentation bug.

I think the `shuffle` doc is indeed wrong here. `array_like` in practice does imply conversion (handwaving definition at https://numpy.org/devdocs/glossary.html#term-array_like), and I can't think of other examples where numpy or scipy functions accept `array_like` and operate on it in-place - that is simply too ill-defined to be a reasonable thing to do.

**Comment by rgommers** [comment]
The fix for this will be included in NumPy 1.20.0, expected to be released in a week. So I'll close this issue. Thanks for the report @cdluminate and for the feedback @rkern.
